pkg/kthena-router/datastore/store.go

```diff
@@ -592,31 +660,109 @@ func (s *store) DeleteModelRoute(namespacedName string) error {
 	return nil
 }
 
-func (s *store) MatchModelServer(model string, req *http.Request) (types.NamespacedName, bool, *aiv1alpha1.ModelRoute, error) {
+func (s *store) MatchModelServer(model string, req *http.Request, gatewayKey string) (types.NamespacedName, bool, *aiv1alpha1.ModelRoute, error) {
 	s.routeMutex.RLock()
 	defer s.routeMutex.RUnlock()
 
 	var isLora bool
-	mr, ok := s.routes[model]
-	if !ok {
-		mr, ok = s.loraRoutes[model]
+	var candidateRoutes []*aiv1alpha1.ModelRoute
+
+	// Try to find routes by model name first
+	routes, ok := s.routes[model]
+	if ok {
+		candidateRoutes = routes
+		isLora = false
+	} else {
+		// Try to find routes by lora name
+		loraRoutes, ok := s.loraRoutes[model]
 		if !ok {
 			return types.NamespacedName{}, false, nil, fmt.Errorf("not found route rules for model %s", model)
 		}
+		candidateRoutes = loraRoutes
 		isLora = true
 	}
 
-	rule, err := s.selectRule(model, req, mr.Spec.Rules)
-	if err != nil {
-		return types.NamespacedName{}, false, nil, fmt.Errorf("failed to select route rule: %v", err)
+	// Try each ModelRoute until we find one that matches
+	for _, mr := range candidateRoutes {
+		// Check parentRefs if specified
+		if len(mr.Spec.ParentRefs) > 0 {
+			// If gatewayKey is provided (not empty), check if ModelRoute matches the specific gateway
+			if gatewayKey != "" {
+				if !s.matchesSpecificGateway(mr, gatewayKey) {
+					continue // Try next ModelRoute
```

Comment:
`One significant motivation to use gateway listeners is to separate same model names' routing. Seems you did not support this at all. To do that you need to store the models with ns+name as the key. When a request from a specific listener comes in, you can know which modelServer can be routed to`

Support here. Only select modelroute with `parentRef` which refering the gateway that listener belong to.